feat(snap): Add snap packaging

farshidtz opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features): https://github.com/edgexfoundry/edgex-docs/pull/790

**If your build fails** due to your commit message not passing the build checks, please review the guidelines here: https://github.com/edgexfoundry/edgex-examples/blob/master/.github/CONTRIBUTING.md.

## What is the current behavior?
<!-- Please describe the current behavior and link to a relevant issue. -->


## Issue Number:


## What is the new behavior?


## Does this PR introduce a breaking change?
<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->

- [ ] Yes
- [x] No

## New Imports
Are there any new imports or modules? If so, what are they used for and why?

## Specific Instructions

Building locally is possible using Snapcraft 7:
```bash
snapcraft  -v
```
You can skip this and install the pre-built snap for this PR.

```bash
# install platform
# We need the 2.2.0+2 which includes the add-on service in secret store
# For older versions add manually: https://docs.edgexfoundry.org/2.2/getting-started/Ch-GettingStartedSnapUsers/#add-on-services
snap install edgexfoundry

# install pre-built snap for this PR
snap install edgex-device-usb-camera --channel=edge/pr-30

# if built manually, connect the edgex-secretstore-token interface to obtain a token
snap connect edgexfoundry:edgex-secretstore-token edgex-device-usb-camera:edgex-secretstore-token

# connect the camera interface :
snap connect edgex-device-usb-camera:camera :camera
```

Configure and start:
```bash
sudo mv /var/snap/edgex-device-usb-camera/current/config/device-usb-camera/res/devices/general.usb.camera.toml.example \
        /var/snap/edgex-device-usb-camera/current/config/device-usb-camera/res/devices/general.usb.camera.toml

# make necessary changes (e.g. the device path)
sudo nano /var/snap/edgex-device-usb-camera/current/config/device-usb-camera/res/devices/general.usb.camera.toml

# start
# this starts both the rstp server and device-usb-camera apps
snap start edgex-device-usb-camera

# get the logs
snap logs -f edgex-device-usb-camera
```

Start streaming:
```bash
curl -X PUT -d '{
    "StartStreaming": {
      "InputImageSize": "640x480",
      "OutputVideoQuality": "5"
    }
}' http://localhost:59882/api/v2/device/name/example-camera/StartStreaming

# check the stream
mplayer rtsp://localhost:8554/stream/example-camera

# stop streaming (to release your camera)
curl -X PUT -d '{}' http://localhost:59882/api/v2/device/name/example-camera/StopStreaming
```

## Other information

The snap is 130MB+. It may be possible to reduce the size by removing unused shared library object files pulled in as ffmpeg dependencies.

farshidtz committed:
feat(snap): initial packaging

Signed-off-by: Farshid Tavakolizadeh <farshid.tavakolizadeh@canonical.com>

farshidtz committed:
feat(snap): fix installation and runtime issues

Signed-off-by: Farshid Tavakolizadeh <farshid.tavakolizadeh@canonical.com>

farshidtz committed:
feat(snap): add interfaces for confinement

Signed-off-by: Farshid Tavakolizadeh <farshid.tavakolizadeh@canonical.com>

farshidtz committed:
fix(snap): move header file to runtime-helpers

Signed-off-by: Farshid Tavakolizadeh <farshid.tavakolizadeh@canonical.com>

farshidtz committed:
fix(snap): add rtsp to auto start, strip binaries

Signed-off-by: Farshid Tavakolizadeh <farshid.tavakolizadeh@canonical.com>

farshidtz committed:
fix: correct location for header file, refactor

Signed-off-by: Farshid Tavakolizadeh <farshid.tavakolizadeh@canonical.com>

farshidtz committed:
docs(snap): clarify auto connection

Signed-off-by: Farshid Tavakolizadeh <farshid.tavakolizadeh@canonical.com>

farshidtz committed:
docs: add general usb camera example

Signed-off-by: Farshid Tavakolizadeh <farshid.tavakolizadeh@canonical.com>

farshidtz committed:
Merge remote-tracking branch 'upstream/main' into snap

tonyespy commented on .gitignore:
```diff
@@ -26,4 +26,14 @@ LintOutput.json
+parts/
+squashfs-root/
+coverage.out
+VERSION
```
Missing NL

tonyespy commented on cmd/res/devices/general.usb.camera.toml.example:
```diff
@@ -0,0 +1,9 @@
+  [DeviceList.Protocols]
+    [DeviceList.Protocols.USB]
+    Path = "/dev/video0"
+    AutoStreaming = "false"
```
Missing NL

I think this is because you merged master into your branch vs. using `git pull -r` to rebase your patches? While your approach works, it means there's more non-related changes to review...

farshidtz replied:
This PR is adding that example.
A clean merge of main into branch does not affect the diff, because the diff is calculated between the branch and main. On the other hand, rebasing will require a re-review of everything because there is no guaranry that the author hasn't changed something else.

farshidtz replied:
Created a separate PR for the example. Will remove from this PR once merged:
- https://github.com/edgexfoundry/device-usb-camera/pull/33

farshidtz replied:
PR has been merged. Removed the example from this PR and merged main into the branch.

tonyespy commented on snap/README.md:
```diff
@@ -0,0 +1,85 @@
+The [snapcraft overview](https://snapcraft.io/docs/snapcraft-overview) provides additional details.
+
+### Obtain a Secret Store token
+<!-- The `edgex-secretstore-token` snap slot makes it possible to automatically receive a token from a locally installed platform snap.-->
```
Why add a hidden comment?

farshidtz replied:
I intend to uncomment it when the auto connection is made possible, i.e. after transferring ownership to Canonical.

tonyespy commented on snap/README.md:
````diff
@@ -0,0 +1,85 @@
+snap start edgexfoundry.security-consul-bootstrapper 
+```
+
+To manually connect and obtain a token:
````
This actually just connects the interface.

In a development scenario, the developer would need to configure the additional settings for the service, then restart the edgexfoundry snap's security-secretstore-setup service.

farshidtz replied:
This is explained above. Will improve the text to make it clearer.

tonyespy commented on snap/README.md:
```diff
@@ -0,0 +1,85 @@
+Please refer [here][secret-store-token] for further information.
+
+### Connect the camera interface
+The `camera` interface is currently not automatically connected. To connect manually:
```
Maybe note why this is needed, namely that the snapd `camera` interface grants permission for the service to access all USB camera on the system.

tonyespy replied:
Also we should submit a request to the store team to grant this interface to the snap automatically.

farshidtz replied:
> Also we should submit a request to the store team to grant this interface to the snap automatically.

Right. That's why I said "currently not automatically connected".
Will add a note to clarify.

farshidtz replied:
Added a link to camera interface docs and note to clarify that auto connection will happen in future.

tonyespy commented on snap/README.md:
```diff
@@ -0,0 +1,85 @@
+# Issues
+The snap is 115 MB, exactly same as the docker image.
+
+It may be possible to reduce the size by removing extra shared library object files.
```
Is there a reason to think that some of the included shared libraries are unused?

farshidtz replied:
This is a draft note (in draft PR). "It may be possible to reduce" and this needs further investigation. Some libraries may not be needed. A statically built go binary is another way to remove reliance on those.

farshidtz replied:
Moved to PR description.

tonyespy commented on snap/local/build-helpers/include/videodev2.h:
```diff
@@ -0,0 +1,2581 @@
+/* SPDX-License-Identifier: ((GPL-2.0+ WITH Linux-syscall-note) OR BSD-3-Clause) */
+/*
+ *  Video for Linux Two header file
```
Why are we copying a Linux kernel header file into the snap build-helpers directory?

```
dpkg -S videodev2.h
linux-hwe-5.13-headers-5.13.0-40: /usr/src/linux-hwe-5.13-headers-5.13.0-40/include/uapi/linux/videodev2.h
linux-hwe-5.13-headers-5.13.0-41: /usr/src/linux-hwe-5.13-headers-5.13.0-41/include/uapi/linux/videodev2.h
linux-hwe-5.11-headers-5.11.0-34: /usr/src/linux-hwe-5.11-headers-5.11.0-34/include/uapi/linux/videodev2.h
linux-hwe-5.13-headers-5.13.0-40: /usr/src/linux-hwe-5.13-headers-5.13.0-40/include/linux/videodev2.h
linux-libc-dev:amd64: /usr/include/linux/videodev2.h
linux-hwe-5.13-headers-5.13.0-41: /usr/src/linux-hwe-5.13-headers-5.13.0-41/include/linux/videodev2.h
linux-hwe-5.11-headers-5.11.0-34: /usr/src/linux-hwe-5.11-headers-5.11.0-34/include/linux/videodev2.h
linux-hwe-5.8-headers-5.8.0-63: /usr/src/linux-hwe-5.8-headers-5.8.0-63/include/linux/videodev2.h
linux-hwe-5.8-headers-5.8.0-63: /usr/src/linux-hwe-5.8-headers-5.8.0-63/include/uapi/linux/videodev2.h
```

**Note** - the above output also shows this header included the dev package for glibc. I honestly don't know enough about v4l2 to say why.

In general, it's not a good idea to copy header files from other projects. If you do, then it's also best practice to be very clear about where the file was sourced from and why. This file is also dual licensed (GPLv2 or BSD), and has some terms about copyright notices that must be included in both **source** and **binary** distributions.

I also ran a diff of a few of the different versions of the header file and they're not all the same across different kernel versions.

This also brings up questions as to whether this device service is agnostic to the underlying kernel version?

farshidtz replied:
I've explained in the snapcraft file why the header is needed.

Good point about the license. However, the header file is included in the resulting snap whether or not I package this version of it. The snap is simply replacing the existing one with a copy that has the set of definitions needed for the compilation.

Regarding being agnostic to the underlying kernel, this probably an issue from the dependency: https://github.com/vladimirvivien/go4vl#compilation-requirements

farshidtz replied:
See issues #37

tonyespy commented on snap/README.md:
```diff
@@ -0,0 +1,85 @@
+If the snap is built and installed locally, the interface will not auto-connect. You can check the status of the connections by running the `snap connections edgex-device-usb-camera` command.
+
+Notes:
+- The auto connection will not happen right now because the snap publisher isn't same as the `edgexfoundry` platrform snap (i.e. Canonical).
```
Is it possible to run this snap in non-secure mode?

MonicaisHer replied:
yes, it is possible, I have tested its streaming function in no-secure mode:
```
snap set edgex-device-usb-camera app-options=true
snap set edgex-device-usb-camera config.edgex-security-secret-store=false
```

tonyespy commented on snap/local/hooks/cmd/install/install.go:
```diff
@@ -0,0 +1,105 @@
+	}
+
+	err = hooks.CopyFile(
+		hooks.Snap+profilesDir+"/general.usb.camera.yaml",
```
Wouldn't it be better to use wildcards? Otherwise, we'd need to update this hook everytime upstream adds new profiles.

farshidtz replied:
All example profiles are copied to the read only snap files and available to the user. The code here is just copying the general one to snap data. 

We may actually be better off not copying any profile/device to snap data because they would be uploaded to core metadata even if not desired. See my other comment: https://github.com/edgexfoundry/device-usb-camera/pull/30#discussion_r879114979

farshidtz replied:
Added code to copy the example device to writable directory. I kept the `.example` suffix to avoid unintentional loading. It isn't needed when dynamic discovery is used.

tonyespy commented on snap/snapcraft.yaml:
```diff
@@ -0,0 +1,134 @@
+name: edgex-device-usb-camera
+base: core20
+license: Apache-2.0
+adopt-info: metadata
```
Do the `summary` and `description` also come from the `metadata` part?

farshidtz replied:
yes

tonyespy commented on snap/snapcraft.yaml:
```diff
@@ -0,0 +1,134 @@
+name: edgex-device-usb-camera
+base: core20
+license: Apache-2.0
```
While the camera service itself is licensed as Apache 2.0, some of the other components in this snap have other licenses (e.g. rtsp-simple-server is MIT).

Also this snap is including ffmpeg which is licensed under LGPLv2, however some components of ffmpeg are licensed under GPLv2 (see https://www.ffmpeg.org/legal.html). Was this reviewed as part of the new service review?

Same question about libpulse0 (which again is licensed under LGPLv2)? The pulseaudio license also mentions that if D-Bus support is enabled, the PulseAudio client library (libpulse) MAY need to be licensed under the GPL, depending on the license adopted for libdbus. Also do we know why service requires libpulse0? Is there a runtime requirement for the PulseAudio server in order to use this service?

farshidtz replied:
> do we know why service requires libpulse0? Is there a runtime requirement for the PulseAudio server in order to use this service?

It is needed by `pprobe`. Added a comment.

Edit: turns out that libpulseaudio it was already there, just that it was missing from path. Removed it as staged package.

farshidtz replied:
> While the camera service itself is licensed as Apache 2.0, some of the other components in this snap have other licenses (e.g. rtsp-simple-server is MIT).

This should be fine since MIT is permissive.
I've also considered not including rtsp server in the snap and let users install their own server (e.g. [v4l2-rtspserver](https://snapcraft.io/v4l2-rtspserver)). But in the end, I decided to keep it since it is only 9MB and disabled by default. The docker image includes it as well.

> Also this snap is including ffmpeg which is licensed under LGPLv2, however some components of ffmpeg are licensed under GPLv2 (see https://www.ffmpeg.org/legal.html). Was this reviewed as part of the new service review?


I did not review the docker base images and packages, not sure if someone else did. @iain-anderson @lenny-intel an idea?
This would be a project issue and not specific to the snap.

jpwhitemn replied:
According to explicit section in GNU's license agreement (section 5 in the document linked below), we are ok to use the library and still distribute as Apache 2.  https://www.gnu.org/licenses/old-licenses/lgpl-2.1.html.  This is known as the linking exception.  The same was applied when we worked with 0MQ.  LF legal cannot provide legal advise (strange but true), but their reading of this also pointed us to this exception.  (see https://en.wikipedia.org/wiki/GPL_linking_exception).

tonyespy replied:
@jpwhitemn you're misinterpreting my comment. I acknowledged that the ffmpeg package is licensed as **LGPLv2**, however there are optional parts of ffmpeg that are licensed as **GPLv2**. My question was simply did we review our usage of ffmpeg to ensure we didn't use one of these optional parts?

iain-anderson replied:
As far as I can tell the ffmpeg support doesn't link the ffmpeg library, it runs an ffmpeg command as a separate process?

tonyespy replied:
My point is that the snap metadata is saying the license of this snap is "Apache 2.0", and that's not technically correct, as the snap (and most likely our docker image) contains more than just the device service Go binary. The guidance from the snapcraft.io [reference](https://snapcraft.io/docs/snapcraft-yaml-reference) for `snapcraft.yaml `say this needs to be an SPDX license expression (see https://spdx.dev/spdx-specification-21-web-version/), which in this case should be a list of licenses as there's doesn't seem to be an option for "multiple licenses" anymore.



For instance, here are a number of shared libraries that and up in the snap and their associated licenses:

- libpango-1.0.0 - LGPL2
- libvorbisenc2 - BSD
- libxvidcore4 - GPL2+

And also while reviewing the files in the snap, I don't see a license file for the rtsp-server itself.

farshidtz replied:
There are close to 200 packages pulled in for ffmpeg and libzmq5. Many of those (incl. libzmq ones) have multiple licenses. It is simply not practical to identify and include all these, and maintain them over time. I don't see any other Docker image or snap listing all licenses.

I found some useful information at:
- https://www.linuxfoundation.org/tools/docker-containers-what-are-the-open-source-licensing-considerations/
- https://opensource.stackexchange.com/a/8393

I believe the license in the Dockerfile is the Dockerfile's license, rather than the license of Docker images built with it. For the snap, the license field appear to be the license of the resulting snap packages. So, IMO, it is most appropriate to remove the optional `license` field. All licenses remain available under snap's `/usr/share/doc` directory.

The license for rtsp-simple-server is at snap's `/usr/share/doc/rtsp-simple-server/LICENSE`.

farshidtz replied:
I have removed the license field from snapcraft.yaml: https://github.com/edgexfoundry/device-usb-camera/pull/30/commits/4c4ab6af09efedeff12eed0a9f83e12c8148dad6

tonyespy commented on snap/snapcraft.yaml:
```diff
@@ -0,0 +1,134 @@
+      # set the version of this snap
+      snapcraftctl set-version $VERSION
+    # parse-info: [edgex-device-usb-camera.metainfo.xml]
+    parse-info: [edgex-device-camera.metainfo.xml]
```
No NL

tonyespy commented on snap/snapcraft.yaml:
```diff
@@ -0,0 +1,134 @@
+    override-build: |
+      cd $SNAPCRAFT_PART_SRC
+
+      # The existing header file misses several of the definitions needed for
```
The existing header file from where? I checked my snapcraft lxc build container and the only `videodev2.h` file I could find is the one you're copying here.

farshidtz replied:
Not sure; probably from core20 (package linux-headers-5.4.0-110).

The go4vl library mentions v5.10.x as the compilation dependency. See https://github.com/vladimirvivien/go4vl#compilation-requirements

farshidtz replied:
Should no longer be an issue when we upgrade to core22.

farshidtz replied:
Switched to core22 and removed the sideloaded header.

tonyespy reviewed (commented):
Please see my inline comments, there are quite a few open questions which need answers.

MonicaisHer commented on snap/README.md:
````diff
@@ -0,0 +1,85 @@
+- The service isn't yet included in the secrets list of the platform snap.
+Add the device service to list of services which have tokens generated for them:
+```
+# Additional secret store tokens
````
```
sudo snap set edgexfoundry app-options=true
```
this is needed

MonicaisHer commented on snap/README.md:
```diff
@@ -0,0 +1,85 @@
+
+# Run the bootstrappers:
+snap start edgexfoundry.security-secretstore-setup
+snap start edgexfoundry.security-consul-bootstrapper 
```
could you add `sudo` for these commands?

farshidtz replied:
Added. Note that this section will be removed once this PR is merged and the snaps released:
- https://github.com/edgexfoundry/edgex-go/pull/4034

farshidtz replied:
Moved to docs:
- https://github.com/edgexfoundry/edgex-docs/pull/772

MonicaisHer commented on snap/local/hooks/cmd/install/install.go:
```diff
@@ -0,0 +1,105 @@
+		return err
+	}
+
+	// no default devices
```
could we add a pre-defined example device into `/var/snap/edgex-device-usb-camera/current/config/device-usb-camera/res/devices/`?

farshidtz replied:
The example device is currently provided under the readonly snap files. The user can copy and modify it if needed:
```
sudo cp /snap/edgex-device-usb-camera/current/config/device-usb-camera/res/devices/general.usb.camera.toml.example \
    /var/snap/edgex-device-usb-camera/current/config/device-usb-camera/res/devices/general.usb.camera.toml
```
I will think about it. Maybe both the example general profile and device can be copied to writable directories but with `.example` suffix. Otherwise, they will get uploaded to core-metadata even if not used or relevant.

farshidtz replied:
Added code to copy the example device to writable directory. I kept the `.example` suffix to avoid unintentional loading. It isn't needed when dynamic discovery is used.

MonicaisHer reviewed (dismissed):
@farshidtz Thanks a lot for the PR. I have tested it with secure and non-secure modes. Both work as expected for video streaming with my camera Aukey-PC-LM1E.

farshidtz committed:
style: add newlines

Signed-off-by: Farshid Tavakolizadeh <farshid.tavakolizadeh@canonical.com>

farshidtz committed:
docs(snap): improve addon service instructions

Signed-off-by: Farshid Tavakolizadeh <farshid.tavakolizadeh@canonical.com>

farshidtz committed:
docs(snap): improve camera interface instrcutionns

Signed-off-by: Farshid Tavakolizadeh <farshid.tavakolizadeh@canonical.com>

farshidtz committed:
docs(snap): add comments to libpulse

Signed-off-by: Farshid Tavakolizadeh <farshid.tavakolizadeh@canonical.com>

farshidtz committed:
docs(snap): move development details to docs and PR description

Signed-off-by: Farshid Tavakolizadeh <farshid.tavakolizadeh@canonical.com>

farshidtz committed:
fix(snap): append libpulse to library paths

Signed-off-by: Farshid Tavakolizadeh <farshid.tavakolizadeh@canonical.com>

farshidtz committed:
docs: Revert "docs: add general usb camera example"

This reverts commit 36ec5bb4839d48288712cbf208a147c93749a790.

Signed-off-by: Farshid Tavakolizadeh <farshid.tavakolizadeh@canonical.com>

farshidtz committed:
feat: switch to core22

- remove sideloaded Ubuntu 22.04 videodev header
- add shared lib paths
- remove libpulse0 as it already exists

Signed-off-by: Farshid Tavakolizadeh <farshid.tavakolizadeh@canonical.com>

farshidtz committed:
fix: install example device config

Signed-off-by: Farshid Tavakolizadeh <farshid.tavakolizadeh@canonical.com>

farshidtz committed:
Merge remote-tracking branch 'upstream/main' into snap

farshidtz committed:
docs: package rtsp-simple-server's licence

Signed-off-by: Farshid Tavakolizadeh <farshid.tavakolizadeh@canonical.com>

farshidtz committed:
docs: add comment to explain the  copy workaround

Signed-off-by: Farshid Tavakolizadeh <farshid.tavakolizadeh@canonical.com>

farshidtz committed:
fix: add missing curl build dependency

Signed-off-by: Farshid Tavakolizadeh <farshid.tavakolizadeh@canonical.com>

MonicaisHer reviewed (dismissed):
@farshidtz  Thanks for the changes. The newest edgex-device-usb-camera snap works as expected.

MonicaisHer commented on snap/snapcraft.yaml:
```diff
@@ -0,0 +1,142 @@
+slots:
+  edgex-secretstore-token:
+    interface: content
+    content: edgex-secretstore-token
```
The `content` can be removed here, as it defaults to the slot name.

farshidtz replied:
Sure. But let's just leave it in for now.

tonyespy commented on snap/local/hooks/cmd/install/install.go:
```diff
@@ -0,0 +1,110 @@
+	}
+
+	err = hooks.CopyFile(
+		hooks.Snap+devicesDir+"/general.usb.camera.toml.example",
```
Why are we copying an example device file when it won't actually be imported unless someone manually renames the file?

farshidtz replied:
This is to maintain consistency among different device services, where configuration files are available in the snap data directories (examples or not). It will be hard for a user to find the example general device if it is not copied over to snap data.

tonyespy reviewed (commented):
I had one comment inline about copying an example device file into` $SNAP_DATA`.

The bigger issue is that I can't seem to build the snap anymore either natively on 22.04 using `--destructive-mode` or using `--use-lxd` on 20.04. In both cases I'm using snapcraft 7.0.2 from `latest/candidate`.

farshidtz committed:
fix: Add scripts to call compiled hooks

This is a workaround for a Snapcraft 7 bug: https://bugs.launchpad.net/snapcraft/+bug/1980089

Signed-off-by: Farshid Tavakolizadeh <farshid.tavakolizadeh@canonical.com>

farshidtz committed:
refactor: upgrade to edgex-snap-hooks and local snap refactoring

Signed-off-by: Farshid Tavakolizadeh <farshid.tavakolizadeh@canonical.com>

MonicaisHer commented on snap/local/helper-go/go.sum:
```diff
@@ -1,5 +1,7 @@
 github.com/canonical/edgex-snap-hooks/v2 v2.2.0 h1:4pDnikrtyrxiynTM49+ppH7hXBx5C7dWUOjEEisCXmI=
 github.com/canonical/edgex-snap-hooks/v2 v2.2.0/go.mod h1:rOxrwdYL7hJDhxFH3uV+nVgLPjWOhJWgM5PRD5YG1jI=
```
`go mod tidy`?

MonicaisHer reviewed (dismissed):
@farshidtz Thanks for the hook refactor. 
I have tested this snap. .env files and config files are in place. Everything works as expected.

farshidtz committed:
build: go mod tidy

Signed-off-by: Farshid Tavakolizadeh <farshid.tavakolizadeh@canonical.com>

farshidtz committed:
test(snap): Add Github workflow for snap testing

Signed-off-by: Farshid Tavakolizadeh <farshid.tavakolizadeh@canonical.com>

farshidtz committed:
fix: write VERSION file to stage directory instead of to source

Needed when building with Snapcraft 7, perhaps as a result of different part sourcing workflow.
This is also a better approach because we won't add files to source files (outside the snap) during the build.

Signed-off-by: Farshid Tavakolizadeh <farshid.tavakolizadeh@canonical.com>

farshidtz committed:
test: switch to v2 of build/test actions

Signed-off-by: Farshid Tavakolizadeh <farshid.tavakolizadeh@canonical.com>

tonyespy commented on snap/README.md:
````diff
@@ -0,0 +1,45 @@
+```
+snapcraft -v
+```
+
````
We should mention that building this snap remains "challenging" and at least offer some options. I don't think the given option "just works".

farshidtz replied:
As explained in [another comment](https://github.com/edgexfoundry/device-usb-camera/pull/30#issuecomment-1175016386), we have been able to build it without problems locally and on CI servers. If there are any OS-specific challenges related to using snapcraft to build core22-based snaps, it should be raised on snapcraft issue trackers.

tonyespy commented on snap/README.md:
```diff
@@ -0,0 +1,45 @@
+Please refer [here][secret-store-token] for further information.
+
+### Connect the camera interface
+The [`camera`](https://snapcraft.io/docs/camera-interface) interface is currently **NOT automatically connected** as it is pending auto-connection permissions from the store.
```
Have we requested these permissions yet?

farshidtz replied:
Not yet. We will do so once we have a stable source and build that can be reviewed by the store.

tonyespy commented on snap/hooks/configure:
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash
+
```
This is a workaround for snapcraft 7, correct? Did you ever get a response from the snapcraft team on this? At minimum, we should probably add a note to this script noting that it's a workaround.

farshidtz replied:
As mentioned in [this comment](https://github.com/edgexfoundry/device-usb-camera/pull/30#issuecomment-1168729209), I've reported the bug.

We are no longer affected by it because we have refactored the hooks to have a single binary executable. This reduces the build complexity (code, time) and snap size. The same improvement is being rolled out to other snaps.

tonyespy commented on snap/hooks/install:
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash -e
+
+exec $SNAP/bin/helper-go install
```
same as configure comment

tonyespy reviewed (changes requested):
While I able to build with snapcraft 6 natively on a 22.04 machine, things still fail for me if I try to build natively using snapcraft 7.0.8 on the same machine. If you want the details, I can provide.

I was able to successfully install the snap, however the testing instructions weren't sufficient for me to actually test the snap.
  - any suggestions on what camera I should use?
  - what are the changes I need to make to the device yaml file?

Also note, there are still some issues with licensing (please see inline comment).

farshidtz committed:
docs: remove license field

License files are packaged under /usr/share/doc

Signed-off-by: Farshid Tavakolizadeh <farshid.tavakolizadeh@canonical.com>

tonyespy reviewed (approved):
Based on our conversation earlier today, I'm approving this PR "as is".
